File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/Variable.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.autodiff.samediff.internal;

import lombok.*;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.autodiff.samediff.VariableType;

import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Data   //TODO immutable?
@Builder
@EqualsAndHashCode(exclude = {"gradient", "variableIndex"})
public class Variable {
    protected String name;
    protected SDVariable variable;
    protected List<String> inputsForOp;
    protected List<String> controlDepsForOp;    //if a op control dependency (x -> opY) exists, then "opY" will be in this list
    protected List<String> controlDepsForVar;   //if a variable control dependency (x -> varY) exists, then "varY" will be in this list
    protected String outputOfOp;        //Null for placeholders/constants. For array type SDVariables, the name of the op it's an output of
    protected List<String> controlDeps;     //Control dependencies: name of ops that must be available before this variable is considered available for execution
    protected SDVariable gradient;      //Variable corresponding to the gradient of this variable
    protected int variableIndex = -1;


    /**
     * Returns true if this variable's array is ready/available for use
     */
    public boolean isArrayReady() {
        SameDiff sameDiff = variable.getSameDiff();
        if (variable.isConstant() || variable.getVariableType() == VariableType.VARIABLE) {
            return sameDiff.arrayAlreadyExistsForVarName(name);
        }
        if (variable.isPlaceHolder()) {
            return sameDiff.arrayAlreadyExistsForVarName(name); // User provided
        }
        if (variable.getVariableType() == VariableType.ARRAY) {
            return sameDiff.arrayAlreadyExistsForVarName(name) ||
                    (sameDiff.isEagerMode() && sameDiff.getEagerArrays().hasArray(name));
        }
        return false;
    }


    /**
     * Returns true if this variable's array can be computed (all inputs are ready)
     */
    public boolean canComputeArray() {
        if (isArrayReady()) return true;
        if (variable.getVariableType() != VariableType.ARRAY) return false;


        SameDiff sameDiff = variable.getSameDiff();
        // Check if all inputs are ready
        if (outputOfOp != null) {
            SameDiffOp op = sameDiff.getOps().get(outputOfOp);
            if (op != null && op.getInputsToOp() != null) {
                for (String input : op.getInputsToOp()) {
                    Variable inputVar = sameDiff.getVariables().get(input);
                    if (inputVar == null || !inputVar.isArrayReady()) {
                        return false;
                    }
                }
                return true;
            }
        }
        return false;
    }

    public List<String> getInputsForOp() {
        return inputsForOp;
    }

    public void setInputsForOp(List<String> inputsForOp) {
        this.inputsForOp = inputsForOp;
    }

    public String getOutputOfOp() {
        return outputOfOp;
    }

    public void setOutputOfOp(String outputOfOp) {
        this.outputOfOp = outputOfOp;
    }
}

File: libnd4j/include/samediff/impl/Variable.cpp
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author raver119@gmail.com
//

#include "samediff/Variable.h"
#include "../SameDiff.h"
#include <samediff/samediff_cpp.h>
#include <NDArrayFactory.h>
#include <exceptions/precondition_exception.h>

namespace samediff {
    Variable::Variable(SameDiff &sd, nd4j::graph::Node *node, int index) {
        precondition_exception::check(node != nullptr, "Variable: Node passed in is null");
        precondition_exception::check(node->parentGraph() != nullptr, "Variable: Node passed in has no Graph defined");
        _node = node;
        _sd = &sd;
        _index = index;
    }

    nd4j::NDArray Variable::array() {
        auto var = _node->parentGraph()->getVariableSpace()->getVariable(_node->id(), _index);
        return *var->getNDArray();
    }

    int Variable::nodeId() const{
        return _node->id();
    }

    SameDiff* Variable::sd() const {
        return _sd;
    }

    Variable Variable::operator+(const Variable& other) const {
        return samediff::arithmetic::Add(*this, other);
    }

    std::pair<int, int> Variable::index() const {
        return {nodeId(), _index};
    }
}

samediff::Variable operator+(const float &scalar, const samediff::Variable &var) {
    auto sd = var.sd();
    auto x = sd->variable(nd4j::NDArrayFactory::create<float>(scalar));
    return samediff::arithmetic::Add(x, var);
}

samediff::Variable operator+(const samediff::Variable &var, const float &scalar) {
    auto sd = var.sd();
    auto y = sd->variable(nd4j::NDArrayFactory::create<float>(scalar));
    return samediff::arithmetic::Add(var, y);
}